generator/integration_tests/golden/internal/golden_kitchen_sink_rest_metadata_decorator.cc

```diff
@@ -0,0 +1,191 @@
+      "routing_id=", {
+      {[](google::test::admin::database::v1::ExplicitRoutingRequest const& request) -> std::string const& {
+        return request.app_profile_id();
+      },
+      std::regex{"profiles/([^/]+)", std::regex::optimize}},
+      {[](google::test::admin::database::v1::ExplicitRoutingRequest const& request) -> std::string const& {
+        return request.app_profile_id();
+      },
+      absl::nullopt},
+      {[](google::test::admin::database::v1::ExplicitRoutingRequest const& request) -> std::string const& {
+        return request.table_name();
+      },
+      std::regex{"(projects/[^/]+)/.*", std::regex::optimize}},
+      }};
+  }();
+  routing_id_matcher->AppendParam(request, params);
+
+  if (params.empty()) {
+    SetMetadata(rest_context);
+  } else {
+    SetMetadata(rest_context, absl::StrJoin(params, "&"));
+  }
+
+  return child_->ExplicitRouting1(rest_context, request);
+}
+
+Status
+GoldenKitchenSinkRestMetadata::ExplicitRouting2(
+    rest_internal::RestContext& rest_context,
+    google::test::admin::database::v1::ExplicitRoutingRequest const& request) {
+  std::vector<std::string> params;
+  params.reserve(2);
+
+  if (!request.app_profile_id().empty()) {
+    params.push_back("no_regex_needed=" + request.app_profile_id());
+  } else if (!request.table_name().empty()) {
+    params.push_back("no_regex_needed=" + request.table_name());
+  } else if (!request.no_regex_needed().empty()) {
+    params.push_back("no_regex_needed=" + request.no_regex_needed());
+  }
+
+  if (!request.nested1().nested2().value().empty()) {
+    params.push_back("routing_id=" + request.nested1().nested2().value());
+  }
+
+  if (params.empty()) {
+    SetMetadata(rest_context);
+  } else {
+    SetMetadata(rest_context, absl::StrJoin(params, "&"));
+  }
+
+  return child_->ExplicitRouting2(rest_context, request);
+}
+
+void GoldenKitchenSinkRestMetadata::SetMetadata(rest_internal::RestContext& rest_context,
+                                        std::string const& request_params) {
+  rest_context.AddHeader("x-goog-request-params", request_params);
+  SetMetadata(rest_context);
+}
+
+void GoldenKitchenSinkRestMetadata::SetMetadata(rest_internal::RestContext& rest_context) {
+  rest_context.AddHeader("x-goog-api-client", api_client_header_);
+  auto const& options = internal::CurrentOptions();
+  if (options.has<UserProjectOption>()) {
+    rest_context.AddHeader(
+        "x-goog-user-project", options.get<UserProjectOption>());
+  }
+  if (options.has<google::cloud::QuotaUserOption>()) {
+    rest_context.AddHeader(
+        "x-goog-quota-user", options.get<google::cloud::QuotaUserOption>());
+  }
+  if (options.has<google::cloud::ServerTimeoutOption>()) {
+    rest_context.AddHeader("x-server-timeout", absl::StrFormat("%.3f",
+      static_cast<double>(options.get<google::cloud::ServerTimeoutOption>().count()) / 1000.0));
```

Comment:
Doesn't `options.get<ServertTimeoutOption>().count() / 1000.0` give you the right `double` without a cast?